File: src/Fuse.Core/Manifests/ServiceManifest.cs
using Fuse.Core.Enums;

namespace Fuse.Core.Manifests
{
    public record ServiceManifest(
        Guid Id,
        string Name,
        string? Version,
        string? Description,
        string? Notes,
        string? Author,
        string? Framework,
        ServiceType Type,
        Uri? RepositoryUri,
        DateTime CreatedAt,
        DateTime UpdatedAt,
        List<DeploymentPipeline> DeploymentPipelines,
        List<Deployments> Deployments,
        List<Tags> Tags
    );

    public record DeploymentPipeline(string Name, Uri? PipelineUri);

    public record Deployments(
        string EnvironmentName,
        Uri? DeploymentUri,
        Uri? SwaggerUri,
        Uri? HealthUri,
        ServiceDeploymentStatus Status
    );


}

File: src/Fuse.Web/src/classes/ServiceManifest.ts
export interface ServiceManifest {
    id: string;
    name: string;
    version: string | null;
    description: string | null;
    notes: string | null;
    author: string | null;
    framework: string | null;
    createdAt: Date | null;
    updatedAt: Date | null;
    repositoryUrl: string | null;
    type: ServiceTypes;
    deploymentPipelines: Array<DeploymentPipeline>;
    deployments: Array<Deployments>;
    tags: Array<string>;
}

interface DeploymentPipeline {
    name: string;
    url: string | null;
}

interface Deployments {
    environment: string;
    url: string | null;
    swaggerUrl: string | null;
    healthUrl: string | null;
    status: DeploymentStatuses;
}

const ServiceTypes = {
    WebApi: 'WebApi',
    Worker: 'Worker',
    FunctionApp: 'FunctionApp',
    Database: 'Database',
    Cache: 'Cache',
    MessageBroker: 'MessageBroker'
} as const;

export type ServiceTypes = typeof ServiceTypes[keyof typeof ServiceTypes];

const DeploymentStatuses = {
    Unknown: 'Unknown',
    Running: 'Running',
    Stopped: 'Stopped',
    Degraded: 'Degraded',
    Maintenance: 'Maintenance',
    Failed: 'Failed'
} as const;

export type DeploymentStatuses = typeof DeploymentStatuses[keyof typeof DeploymentStatuses];

export type ServiceManifestCreate = Omit<
	ServiceManifest,
	"id" | "createdAt" | "updatedAt"
> & {
	id?: string;
	createdAt?: Date | string | null;
	updatedAt?: Date | string | null;
};